File: ios/Runner/Callkit/Models/Call.swift
//
//  Call 2.swift
//  Runner
//
//  Created by Hashim Khan on 24/01/2024.
//


import Foundation
import AVFAudio

public class Call: NSObject , ObservableObject {
    
    @Published var uuid: UUID
    @Published var callPayload : CallPayload
    @Published var isOutGoing: Bool
    
    public var handle: String?
    var connectingDate: Date?
    var connectedDate: Date?
    var endDate: Date?
    var isOnHold = false
    var isMuted = false
    
    var hasStartedConnecting: Bool{
        get{
            return connectingDate != nil
        }
        set{
            connectingDate = newValue ? Date() : nil
        }
    }
    
    var hasConnected: Bool {
        get{
            return connectedDate != nil
        }
        set{
            connectedDate = newValue ? Date() : nil
        }
    }
    
    init(uuid: UUID, callPayload: CallPayload, isOutGoing: Bool = false){
        self.uuid = uuid
        self.callPayload = callPayload
        self.isOutGoing = isOutGoing
    }
    
    
    func callStarted(){
        guard hasStartedConnecting == false else { return }
        hasStartedConnecting = true
    }
        
    func callAccepted(){
        guard hasStartedConnecting == false else { return }
        hasStartedConnecting = true
    }
        
    func callConnected(){
        guard hasConnected == false else { return }
        hasConnected = true
    }
}

File: android/app/src/main/kotlin/com/transport_system/ts_driver/telecom/models/Call.kt
package com.transport_system.ts_driver.telecom.models

import java.util.Date
import java.util.UUID

class Call(
    var uuid: UUID,
    var callPayload: CallPayload,
    var isOutGoing: Boolean
) {

    private var connectingDate: Date? = null
    private var connectedDate: Date? = null



    var hasStartedConnecting: Boolean
        get() = connectingDate != null
        set(value) {
            connectingDate = if (value) Date() else null
        }


    var hasConnected: Boolean
        get() = connectedDate != null
        set(value) {
            connectedDate = if (value) Date() else null
        }


    fun callStarted() {
        if (!hasStartedConnecting) {
            hasStartedConnecting = true
        }
    }

    fun callAccepted() {
        if (!hasStartedConnecting) {
            hasStartedConnecting = true
        }
    }

    fun callConnected() {
        if (!hasConnected) {
            hasConnected = true
        }
    }

    // Match by UUID when the payload has a usable one, else by conversationId (only one call is active).
    fun matches(payload: CallPayload): Boolean {
        val payloadUuid = try {
            payload.tempCallId?.let { UUID.fromString(it) }
        } catch (_: Exception) {
            null
        }
        if (payloadUuid != null && payloadUuid == uuid) return true
        val convoId = callPayload.conversationId
        return convoId != null && convoId == payload.conversationId
    }

}